# Regresión lineal simple
### División de los datos en set de entrenamiento y validación

In [1]:
import numpy as np # https://numpy.org/
import pandas as pd # https://pandas.pydata.org/
import seaborn as sns # https://seaborn.pydata.org/
import matplotlib.pyplot as plt # https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.html
%matplotlib inline

In [2]:
url = 'https://raw.githubusercontent.com/JASDataCTG/Diplomado-ML/main/Modulo%202/Datasets/precio_prep_noparq.csv'
df = pd.read_csv(url, header = 0) # Leer el archivo generado en el notebook de preproceso

In [3]:
# Crear un nuevo dataframe sin la columna precio (axis=1 -> borra columna)
X_multiple = df.drop('precio', axis = 1)

In [4]:
y_multiple = df['precio']

In [5]:
# Importar el modulo y la función necesaria para el modelo de regresión lineal
from sklearn.linear_model import LinearRegression

# Importar modulo y función necesaria para hacer una división aleatoria de los datos
from sklearn.model_selection import train_test_split 

In [6]:
# Obtener los cuatro conjuntos de datos necesarios para el modelo
X_train, X_test, y_train, y_test = train_test_split(X_multiple, y_multiple, test_size = 0.2, random_state = 123)
#  test_size: 0.2 es el 20 % para que sean datos de validacion
#  random_state : utiliza la semilla igual para todo las ejecuciones

In [7]:
X_train.shape, X_test.shape, y_test.shape, y_train.shape # Revisar las dimensiones de las matrices de los dataframes generados

((404, 15), (102, 15), (102,), (404,))

Las 506 instancias del conjunto de datos se dividen en un set de entrenamiento de 404 instancias y otro de validación con 102 instancias, equivalente al 80 y 20% respectivamente. Se asigna una semilla para generar los datos con el fin de que este sea reproducible y no cambien los datos en cada ejecución de la celda del notebook.

In [10]:
linear_model_s = LinearRegression()

In [11]:
linear_model_s.fit(X_train, y_train)
# ajuste a los datos

LinearRegression()

In [12]:
# Realizar la predicción en el conjunto de entrenamiento
y_train_s_pred = linear_model_s.predict(X_train)

In [13]:
# Realizar la predicción en el conjunto de validación
y_test_s_pred = linear_model_s.predict(X_test)

In [14]:
# Ver las predicciones desarrolladas por el modelo para el conjunto de validación X_test
y_test_s_pred

array([21.85042285, 28.79400287, 37.90874663, 17.99761025, 30.83149815,
       35.43633019, 27.04032829, 11.93328437, 12.22342317, 30.26511201,
       28.45219631, 19.42504799, 16.06842444, 32.0297096 , 16.14155676,
       20.72201702, 21.79088247, 35.88044494, 18.4247318 , 13.58032732,
       13.87030054, 24.58137746, 33.15916941, 38.7284764 , 29.67295715,
       20.94152244, 16.16407564, 19.81264505, 25.04346478, 21.71070371,
       27.06003084, 33.45409273, 18.98131883, 21.349003  , 23.87355846,
       33.65982708, 27.60366996, 15.96767857, 14.98534173, 37.16042019,
       31.1500621 , 20.31380503, 18.97497791, 40.84918794,  6.41062112,
       29.77874167, 20.84633752, 27.04229197, 16.3085342 , 26.12120801,
       32.69007448, 24.42849945, 20.47464687,  9.68017703, 25.45276451,
       17.95587437, 10.28964633,  0.19337747, 29.75796808,  9.81774259,
       21.11217566, 21.70352473, 22.08564583, 17.67456293, 21.42157936,
       24.06423738, 24.06989918, 24.00197505, 23.84660286, 31.33

In [15]:
# Ver el conjunto de validación
X_test

tasa_crimen  area_resid  cal_aire  num_hab   edad  prof_mil  p_pobreza  \
410     3.953852       48.10    0.5970    5.757  100.0      19.8      10.11   
85      0.055766       34.49    0.4490    6.630   56.1      21.5       6.53   
280     0.035155       33.33    0.4429    7.820   64.5      25.1       3.76   
422     2.568650       48.10    0.6140    5.648   87.6      19.8      14.10   
199     0.031014       31.47    0.4030    6.975   15.3      23.0       4.56   
..           ...         ...       ...      ...    ...       ...        ...   
229     0.365878       36.20    0.5040    6.552   21.4      22.6       3.76   
159     0.885840       49.58    0.8710    6.510  100.0      25.3       7.39   
196     0.039326       31.52    0.4040    7.287   34.1      27.4       4.08   
345     0.030655       34.39    0.4420    6.014   48.5      21.2      10.53   
467     1.690516       48.10    0.5840    6.003   94.5      19.8      21.32   

     n_cam_hos  n_hab_hotel  lluvia_ano  dist_prom  aeropuerto_SI  \
410      9.600      11.1200          52     1.4125              0   
85       9.732      12.2128          56     4.4400              0   
280      7.108      10.3632          50     4.6950              0   
422      8.916      11.1664          36     1.9525              0   
199      7.998      13.2792          24     7.6525              1   
..         ...          ...         ...        ...            ...   
229      7.330      15.2520          47     3.3725              1   
159      7.166      12.1864          36     1.7675              1   
196      7.266      15.2664          20     7.3100              1   
345      7.150      13.1400          26     8.0125              1   
467      9.482      13.1528          21     2.5400              0   

     cuerpo_agua_Lago  cuerpo_agua_Lago y Rio  cuerpo_agua_Rio  
410                 0                       0                1  
85                  0                       0                1  
280                 0                       0                0  
422                 0                       0                0  
199                 0                       0                0  
..                ...                     ...              ...  
229                 0                       0                0  
159                 1                       0                0  
196                 0                       1                0  
345                 1                       0                0  
467                 0                       1                0  

[102 rows x 15 columns]

In [17]:
# Convertir el array vector de predicciones a dataframe
y_test_s_df = pd.DataFrame(y_test_s_pred, columns = ['Prediccion'])
## Se agrega la columna de prediccion
## y_test_s_pred es un array y lo conviente en un dataframe

In [18]:
y_test_s_df

Prediccion
0     21.850423
1     28.794003
2     37.908747
3     17.997610
4     30.831498
..          ...
97    32.282488
98    28.248453
99    35.679727
100   19.594359
101   14.146660

[102 rows x 1 columns]

In [19]:
# Reindexar X_texst para poder hacer el join con y_test_s_df
X_test_reindex = X_test.reset_index(drop = True, inplace = False)
## Si hago join del df con las caracteristicas X, saldria error porque el index esta mal. con este comando se reinicia el index de x

In [20]:
# Realizar la fusión de los dos datafram para poder ver las variables predictoras
# y la variable predica en un solo dataframe
test_x_con_pred = X_test_reindex.join(y_test_s_df)

In [21]:
test_x_con_pred

tasa_crimen  area_resid  cal_aire  num_hab   edad  prof_mil  p_pobreza  \
0       3.953852       48.10    0.5970    5.757  100.0      19.8      10.11   
1       0.055766       34.49    0.4490    6.630   56.1      21.5       6.53   
2       0.035155       33.33    0.4429    7.820   64.5      25.1       3.76   
3       2.568650       48.10    0.6140    5.648   87.6      19.8      14.10   
4       0.031014       31.47    0.4030    6.975   15.3      23.0       4.56   
..           ...         ...       ...      ...    ...       ...        ...   
97      0.365878       36.20    0.5040    6.552   21.4      22.6       3.76   
98      0.885840       49.58    0.8710    6.510  100.0      25.3       7.39   
99      0.039326       31.52    0.4040    7.287   34.1      27.4       4.08   
100     0.030655       34.39    0.4420    6.014   48.5      21.2      10.53   
101     1.690516       48.10    0.5840    6.003   94.5      19.8      21.32   

     n_cam_hos  n_hab_hotel  lluvia_ano  dist_prom  aeropuerto_SI  \
0        9.600      11.1200          52     1.4125              0   
1        9.732      12.2128          56     4.4400              0   
2        7.108      10.3632          50     4.6950              0   
3        8.916      11.1664          36     1.9525              0   
4        7.998      13.2792          24     7.6525              1   
..         ...          ...         ...        ...            ...   
97       7.330      15.2520          47     3.3725              1   
98       7.166      12.1864          36     1.7675              1   
99       7.266      15.2664          20     7.3100              1   
100      7.150      13.1400          26     8.0125              1   
101      9.482      13.1528          21     2.5400              0   

     cuerpo_agua_Lago  cuerpo_agua_Lago y Rio  cuerpo_agua_Rio  Prediccion  
0                   0                       0                1   21.850423  
1                   0                       0                1   28.794003  
2                   0                       0                0   37.908747  
3                   0                       0                0   17.997610  
4                   0                       0                0   30.831498  
..                ...                     ...              ...         ...  
97                  0                       0                0   32.282488  
98                  1                       0                0   28.248453  
99                  0                       1                0   35.679727  
100                 1                       0                0   19.594359  
101                 0                       1                0   14.146660  

[102 rows x 16 columns]

## Validación del modelo

In [22]:
# Importar la función que realiza el calculo del coeficiente de determinación
# con el fin de analizar la bondad de ajuste del modelo
from sklearn.metrics import r2_score

In [23]:
r2_score?

Signature:
r2_score(
    y_true,
    y_pred,
    *,
    sample_weight=None,
    multioutput='uniform_average',
)
Docstring:
:math:`R^2` (coefficient of determination) regression score function.

Best possible score is 1.0 and it can be negative (because the
model can be arbitrarily worse). A constant model that always
predicts the expected value of y, disregarding the input features,
would get a :math:`R^2` score of 0.0.

Read more in the :ref:`User Guide <r2_score>`.

Parameters
----------
y_true : array-like of shape (n_samples,) or (n_samples, n_outputs)
    Ground truth (correct) target values.

y_pred : array-like of shape (n_samples,) or (n_samples, n_outputs)
    Estimated target values.

sample_weight : array-like of shape (n_samples,), default=None
    Sample weights.

multioutput : {'raw_values', 'uniform_average', 'variance_weighted'},             array-like of shape (n_outputs,) or None, default='uniform_average'

    Defines aggregating of multiple output scores.
    Array

In [24]:
# Ver la bondad de ajuste en los datos de entrenamiento
r2_score(y_train, y_train_s_pred)

0.7476776519461668

El valor de $r^2$ del modelo nos informa que el modelo es capaz de explicar, aproximadamente, el 25% de la varianza de la variable predicha de los precios de las propiedades en el conjunto de entrenamiento.

_*Nota25:*_ sale de 1-0.7476, que es el error.
el 74.76% se puede explicar con el modelo
cuando es 1 el $r^2$, esta sobre entrenado, o sobre ajustado, lo que no permite predecir a nuevos datos

In [25]:
r2_score(y_test, y_test_s_pred) # Ver la bondad de ajuste en los datos de validación

0.5948163371697073

El valor de $r^2$ del modelo nos informa que el modelo es capaz de explicar, aproximadamente, el 40.5% de la varianza de la variable predicha de los precios de las propiedades en el conjunto de validación.

### Validación cruzada (KFold)
![Validación cruzada](https://github.com/JASDataCTG/Diplomado-ML/blob/main/Modulo%202/Notebooks/5-Fold-Cross-Validation.jpg?raw=1)


El método de validación que vimos en la sección anterior se conoce como el método de validación simple, muy utilizado para conjuntos de tamaño considerable (más de 2 millones de instancias), sin embargo, cuando tenemos datasets con pocas instancias (más de 200 y menos de 2 millones) se puede utilizar una técnica conocida como *Cross Validation* o validación cruzada. La validación cruzada consiste en dividir los datos de forma aleatoria en $k$ grupos de aproximadamente el mismo tamaño, $k-1$ grupos se emplean para desarrollar el entremiento del modelo y uno de los grupos se emplea para la validación.

In [26]:
# Importar el modulo necesario para desarrollar la validación cruzada
from sklearn.model_selection import KFold

# Importar las métricas de evaluación del modelo (score)
from sklearn.model_selection import cross_val_score

In [27]:
# Escoger las porciones en que vamos a dividir el dataset de entrenamiento para hacer las particiones y las pruebas
kf = KFold(n_splits = 10)
## 10 son los folds

In [28]:
# Hallar el valor de r2 con la función score
score = linear_model_s.score(X_train, y_train)
## esta es otra forma de calcular el r2

In [29]:
score

0.7476776519461668

In [31]:
# Desarrollar la validación cruzada con el conjunto de datos de entrenamiento
score_cv_train = cross_val_score(linear_model_s, X_train, y_train, cv = kf)

In [32]:
score_cv_train
## me devuelve el R2 para cada conjunto de los 10 folds

array([0.56618134, 0.71670555, 0.82868053, 0.83212899, 0.64932695,
       0.84655635, 0.57217345, 0.76415869, 0.49235989, 0.71077861])

In [33]:
# Calcular la media de los score (r cuadrado) de las diez validaciones desarrolladas
print("Media de las validaciones set de entrenamiento: ", score_cv_train.mean())

Media de las validaciones set de entrenamiento:  0.6979050335940291


Observar que con la validación cruzada se obtiene un valor de $r^2$ menor al obtenido por la validación simple o particionada.

## Realizar el score para el set de validación de datos de validación

In [34]:
score_cv_train_cross = cross_val_score(linear_model_s,X_test, y_test, cv=kf)

In [35]:
score_cv_train_cross

array([0.61540506, 0.19474075, 0.40707726, 0.69401953, 0.51736522,
       0.47329125, 0.26056727, 0.33555776, 0.60323547, 0.13524216])

In [36]:
print("Media ", score_cv_train_cross.mean())
## observe que mejora con relacion al uso sin fold

Media  0.42365017427012425
